src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,343 @@
+    // guard goes out of scope here
+}
+```
+
+When `guard` goes out of scope, it will block execution until the thread is
+finished. If we didn't want this behaviour, we could use `Thread::spawn()`:
+
+```
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    Thread::spawn(|| {
+        println!("Hello from a thread!");
+    });
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+We need to `sleep` here because when `main()` ends, it kills all of the
+running threads.
+
+[`scoped`](std/thread/struct.Builder.html#method.scoped) has an interesting
+type signature:
+
+```text
+fn scoped<'a, T, F>(self, f: F) -> JoinGuard<'a, T>
+    where T: Send + 'a,
+          F: FnOnce() -> T,
+          F: Send + 'a
+```
+
+Specifically, `F`, the closure that we pass to execute in the new thread.
+It has two restrictions: It must be a `FnOnce` from `()` to `T`. This allows
+us to take ownership of the new thread's environment. The other restriction is
+that `F` must be `Send`. We aren't allowed to transfer this ownership unless
+the type thinks that's okay.
+
+Many languages have the ability to execute threads, but it's wildly unsafe.
+There are entire books about how to prevent errors that occur from shared
+mutable state. Rust helps out with its type system here, as well. Let's talk
+about how you actually share things between threads.
+
+## Safe Shared Mutable State
+
+Due to Rust's type system, we have a concept that sounds like a lie: "safe
+shared mutable state." Many programmers agree that shared mutable state is
+very, very bad.
+
+Someone once said this:
+
+> Shared mutable state is the root of all evil. Most languages attempt to deal
+> with this problem through the 'mutable' part, but Rust deals with it by
+> solving the 'shared' part.
+
+The same [ownership system](ownership.html) that helps prevent data races in
+memory can help us make the same guarantees about shared references to state
+in our concurrent programs.
+
+As an example, here is a Rust program that would have a data race in many
+languages. It will not compile:
+
+```ignore
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    let mut data = vec![1u32, 2, 3];
+
+    for i in 0 .. 2 {
+        Thread::spawn(move || {
+            data[i] += 1;
+        });
+    }
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+This gives us an error:
+
+```text
+12:17 error: capture of moved value: `data`
+        data[i] += 1;
+        ^~~~
+```
+
+In this case, we know that our code _should_ be safe, but Rust isn't sure. And
+it's actually not safe: if we had a reference to `data` in each thread, and the
+thread takes ownership of the reference, we have three owners! That's bad. We
+can fix this by using the `Arc<T>` type, which provides an atmoic reference
+count. The 'atomic' part means that it's safe to share across threads.
+
+Here's the second version of our code. It still doesn't work, but for a
```

Comment:
This is a confusing transition -- above you mention `Arc`, but in the revised example you're not using it.